### DISCUSSION TITLE
Plugins and WholeBodyDynamics in Simulation

### DISCUSSION THREAD

**Discussion by diogofbsilva** [robotology/community#697] [Q&A] [answered]
Hello everyone!

I am working on my thesis in ISR Lisbon with the iCub where I have the goal to control the dexterous robot hands using tactile and force sensors. In order to have good testing I need to simulate grasping activities. The work done here https://github.com/robotology/icub-gazebo-grasping-sandbox/tree/master creates a grasping activity. which is ideal to my work, being that the reason that I am using it.  As I mentioned earlier, in this thesis I am using tactile and FT sensors so the use of plugins is important. However, as of right now, in my simulations, I do not have ports that are active giving FT readings. 

So, here is my first question: what might be the problem I am facing, provided that I have in the robotology pipeline the force torque plugins.

The second problem is regarding the use of the wholeBodyDynamics device in simulation as well. Do you have any pointers that may help me using this device in the simulation? In this device, calibration of the FT sensors is made which is very helpful to my thesis to not take into account the weight of the arm in the sensor measurements.

Thank you
Diogo Silva

**Comment by pattacini** [comment]
Hi @diogofbsilva 

The sandbox makes use of the model [`iCubGazeboV2_5_visuomanip`](https://github.com/robotology/icub-gazebo-grasping-sandbox/blob/master/gazebo/worlds/icub-gazebo-grasping-sandbox.sdf#L57), which has been specifically designed to focus on the hands but, on the other side, it does not expose FT sensors.

Therefore, you would need to extend the model taking inspiration from a more "standard" one.

Here's a list of useful pointers:
- https://github.com/robotology/icub-models/blob/836440918c72a0484eff069ef03d61dca3a3eeb6/iCub/robots/iCubGazeboV2_5/model.urdf#L2650-L2689
- https://github.com/robotology/icub-models/blob/836440918c72a0484eff069ef03d61dca3a3eeb6/iCub/conf/icub.xml#L22-L31

For the `wholebodydynamics`, our warm recommendation is to rely on the idyntree-based service.
Take a look at how it's called for a physical robot:
- https://github.com/robotology/robots-configuration/blob/c6a298257827bb08de3c6b734ea7144399531df8/ergoCubSN001/ergocub_wbd.xml#L141

CC'ing @robotology/iit-artificial-mechanical-intelligence to get better pointers for how to configure the WBD service on Gazebo.

**Reply by diogofbsilva** [reply]
Hello!
Thank you for the very useful pointers, especially the ones regarding the FT sensors as I was able to have them working on my simulation. Thank you also for the pointers regarding the WBD and for CC'ing more people to help me! I really apprecciate it!

**Comment by traversaro** [comment]
> The second problem is regarding the use of the wholeBodyDynamics device in simulation as well. Do you have any pointers that may help me using this device in the simulation? In this device, calibration of the FT sensors is made which is very helpful to my thesis to not take into account the weight of the arm in the sensor measurements.

So, in simulation, the FT sensors already provide you the correct measurement without any offset, so there is no need to calibrate an additional offset. To disable the calibration of offset, you can pass the `startWithZeroFTSensorOffsets` option, see the `wholebodydynamics` device documentation (https://github.com/robotology/whole-body-estimators/tree/v0.10.0/devices/wholeBodyDynamics) or https://github.com/robotology/whole-body-estimators/pull/72 and https://github.com/robotology/whole-body-estimators/issues/67 for more details.

Then, in simulation you can either run the wbd externally, or run it as part of the robotinterface of the simulated robot itself (see the `wbd-ext` and `wbd-int` use cases described in https://github.com/icub-tech-iit/ergocub-software/issues/35). At the moment in `ergocub-software` we are using the `wbd-ext` workflow, see https://github.com/icub-tech-iit/ergocub-software/blob/v0.7.3/urdf/ergoCub/conf/launch_wholebodydynamics_ecub.xml and https://github.com/icub-tech-iit/ergocub-software/blob/v0.7.3/urdf/ergoCub/conf/estimators/wbd_ecub_sim.xml for the related configuration files. Similar configuration files for simulated iCub are https://github.com/robotology/whole-body-estimators/blob/master/devices/wholeBodyDynamics/app/wholebodydynamics-icub-external-six-fts-sim.xml and https://github.com/robotology/whole-body-estimators/blob/master/devices/wholeBodyDynamics/app/launch-wholebodydynamics-icub-six-fts-sim.xml .

**Reply by diogofbsilva** [reply]
Hello @traversaro! Thank you so much for your help!

Yes, in simulation the calibration of an offset is not needed, however I have another doubt for you.
Is in simulation the computations done in https://github.com/robotology/icub-main/blob/9f14d25946fdf22e8efc342f23ff781af6def5b0/src/modules/wholeBodyDynamics/observerThread.cpp#L841-L842 still useful?
Sorry for sending a link of the WBD in the iCub-main however I navigate better in the module than in the device for now.

As for how to run the wbd in simulation I will look into what sent and if I get into any trouble I will get in contact.

Thank you so much for your help!!
Happy Sunday!

**Comment by diogofbsilva** [comment]
Hello @traversaro @pattacini !

For the past few weeks I have been working with the FT plugins as they are working well in my simulation, however the information that is online is confusing me a little bit as the axes of the FT sensors of the image below are not matching with the results I am obtaining:

![kinematics](https://github.com/robotology/community/assets/102809063/579c0e07-4123-44c9-b873-d21a3bddef32)

I am saying this because when analyzing the force values for different positions of the left arm of the iCub the values did not make sense to me taking into account the axes of the image. Take for example, when the arm of the iCub is up: It would be expected that for this position |Fx| would be the biggest, however that does not happen (|Fz| is the biggest value):

![image](https://github.com/robotology/community/assets/102809063/f9147064-6c0c-4a07-884b-f8a38324a6b6)
![Fx_up](https://github.com/robotology/community/assets/102809063/3c75b9d7-3e00-43f5-b134-7131560e2747)
![Fz_up](https://github.com/robotology/community/assets/102809063/c5666629-601a-4b7e-a78f-a8b1c1cc211d)

When analyzing the force values when the arm is down it seems that the same thing happens (the axis x seems switched with the axis z), as the force |Fz| should be the bigger value:

![image (1)](https://github.com/robotology/community/assets/102809063/a0a597cd-0a0d-4964-bbd2-29b9aeb2fec2)
![Fx_down](https://github.com/robotology/community/assets/102809063/dc32c583-3837-4513-a913-b14ac6e40e1a)
![Fz_down](https://github.com/robotology/community/assets/102809063/248bd827-16b0-4c56-9b5b-857fbed7b791)


Are the axis represented in the first image not correct? What seems to be my problem?

Thank you for your help!!

Diogo

**Reply by Nicogene** [reply, marked as answer]
Hi @diogofbsilva,

We are aware that the FT sensors in iCub URDF have problems:
- https://github.com/robotology/icub-models-generator/issues/233
- https://github.com/robotology/icub-models-generator/issues/271

this could explain the inconsistency you are seeing, we will let you know once we find a solution to this problem

dd @traversaro @pattacini @martinaxgloria

**Reply by traversaro** [reply]
@diogofbsilva which models are you using in your simulation?

**Reply by diogofbsilva** [reply]
Hi @traversaro !

I am using the model iCubGazeboV2_5_visuomanip, with slight modifications in the URDF such as the addition of the FT sensors located in the arms, which I got from the URDF model of iCubGazeboV2_5.

**Reply by diogofbsilva** [reply]
@Nicogene thank you for your help!
You have come up with a fix that can be quite helpful for me. https://github.com/robotology/icub-models-generator/issues/233#issuecomment-1997548531
 I will look into it!

Thanks again!

**Reply by traversaro** [reply]
Ok, thanks for the details! If you need further help, probably it is useful if you push somewhere the model you are using if it is not one of the one availably by default in icub-models .
